src/transformers/vocabulary2id.py

```diff
@@ -0,0 +1,81 @@
+import operator
+
+from pyspark import RDD, Row
+from models.code2vec_features import Code2VecFeatures
+
+from ast import literal_eval as make_tuple
+from sourced.ml.transformers import Transformer
+
+
+class Vocabulary2Id(Transformer):
+    def __init__(self, keep_freq: bool, output: str, **kwargs):
+        super().__init__(**kwargs)
+        self.keep_freq = keep_freq
+        self.output = output
+
+    def __call__(self, rows: RDD):
+        self.value2index, self.path2index = self.build_vocabularies(rows)
+
+        doc2path_contexts = self.build_doc2pc(rows)
+
+        doc2path_contexts = doc2path_contexts.collect()
+
+        Code2VecFeatures().construct(value2index=self.value2index, path2index=self.path2index,
+                                     path_contexts=doc2path_contexts).save(
+            self.output)
+
+    @staticmethod
+    def _unstringify_path_context(row):
+        """
+        Takes a row containing ((pc, doc), freq) and returns a tuple (u, path, v)
+        (removes namespace prefix v.)
+        """
+        return make_tuple(row[0][0][2:])
+
+    def _get_path(self, row: Row):
+        # 2: removes the namespace v. from the string to parse it as tuple
+        k = Vocabulary2Id._unstringify_path_context(row)
+        return [(k[0], 1), (k[1], 1), (k[2], 1)]
+
+    def build_vocabularies(self, rows: RDD):
+        """
+        Process rows to gather values and paths with/without their frequencies.
+        """
+
+        rows = rows \
+            .flatMap(self._get_path) \
+            .reduceByKey(operator.add) \
+            .persist()
+
+        get_elem = lambda x: x[0]
+        if not self.keep_freq:
+            rows = rows.map(lambda x: x[0])
+            get_elem = lambda x: x
+
+        values = rows.filter(lambda x: type(get_elem(x)) == str).collect()
+        paths = rows.filter(lambda x: type(get_elem(x)) == tuple).collect()
+
+        value2index = {w: i for i, w in enumerate(values)}
```

Comment:
this is a problem if `keep_freq is True`, in one case you get the dictionnaries but in the other you get `(key, value): index`. I would move the index creation to the call, and use `self.docfreq` to potentially create the frequency dict, and pass it as an optional dict parameter for the Model that default to None.